contracts/Strategy.sol

```diff
@@ -53,80 +222,624 @@ contract Strategy is BaseStrategy {
+                    lowestInterestPairIndex(
+                        Math.min(
+                            sharesToFreeFromKashi,
+                            wantToBentoShares(
+                                estimatedTotalAssets().div(kashiPairs.length),
+                                true
+                            )
+                        )
+                    );
+                if (lowestInterestIndex != type(uint256).max) {
+                    sharesFreedFromKashi = liquidateKashiPair(
+                        lowestInterestIndex,
+                        sharesToFreeFromKashi
+                    );
+                }
 
-        uint256 totalAssets = want.balanceOf(address(this));
-        if (_amountNeeded > totalAssets) {
-            _liquidatedAmount = totalAssets;
-            _loss = _amountNeeded.sub(totalAssets);
+                for (
+                    uint256 i = 0;
+                    i < kashiPairs.length &&
+                        sharesToFreeFromKashi > sharesFreedFromKashi;
+                    i++
+                ) {
+                    if (i == lowestInterestIndex) continue; // we already visited this
+
+                    sharesFreedFromKashi = sharesFreedFromKashi.add(
+                        liquidateKashiPair(
+                            i,
+                            sharesToFreeFromKashi.sub(sharesFreedFromKashi)
+                        )
+                    );
+                }
+
+                bentoBox.withdraw(
+                    BIERC20(address(want)),
+                    address(this),
+                    address(this),
+                    0,
+                    sharesInBento()
+                );
+            }
+
+            _liquidatedAmount = balanceOfWant();
+
+            if (_amountNeeded > _liquidatedAmount) {
+                _loss = _amountNeeded.sub(_liquidatedAmount);
+            }
         } else {
             _liquidatedAmount = _amountNeeded;
         }
     }
 
-    function liquidateAllPositions() internal override returns (uint256) {
-        // TODO: Liquidate all positions and return the amount freed.
+    function liquidateAllPositions()
+        internal
+        override
+        returns (uint256 _liquidatedAmount)
+    {
+        (_liquidatedAmount, ) = liquidatePosition(type(uint256).max);
+    }
+
+    // new strategy **must** have the same kashiPairs attached
+    function prepareMigration(address _newStrategy) internal override {
+        for (uint256 i = 0; i < kashiPairs.length; i++) {
+            if (kashiPairs[i].pid != 0) {
+                masterChef.withdraw(
+                    kashiPairs[i].pid,
+                    kashiFactionInMasterChef(i)
+                );
+            }
+
+            kashiPairs[i].kashiPair.transfer(
+                _newStrategy,
+                kashiFractionInPair(i)
+            );
+        }
+    }
+
+    function addKashiPair(address _newKashiPair, uint256 _newPid)
```

Comment:
Can you expand when a pid will be 0 vrs not? i see special logic below for being different than zero

Reply:
@storming0x I'll add some comments, but a non-zero pid indicates that there are masterchef staking/rewards available. While 0 is a valid pid, we know for kashiPairs that the pid will always be greater than 0, so we are using that as a sentinel value.

Reply:
To be clear
- pid == 0
  - no masterchef staking
- pid > 0
  - stake in masterchef
  - collect rewards